nengo/builder/signal.py

```diff
@@ -259,6 +260,92 @@ def row(self):
         return self.reshape((1, self.size))
 
 
+class SparseSignal:
```

Comment:
I agree that that would be nice. Just be aware that it's a reasonably big undertaking. We make assumptions in a few places about how signals are done internally, and I worry that backends are making more assumptions.

The reason I added the `SparseDotInc` operator was so that this could explicitly break on backends (though frankly, for NengoOCL, it might try to allocate data for the SparseSignals first and break in a weird way because of that, so I don't think that's the best way to do it). It would be nice if whatever way we do it with Signals, it could be obvious to backends what type of signal it is, and they can immediately error if they don't support that. Originally, I was imagining an abstract `Signal` base class that then gets implemented as e.g. `DenseSignal`, `SparseSignal`. But there are definitely some differences between the APIs (e.g. the dense version has `offset`, `strides`, etc., which are not applicable to the sparse one, though we could just return dummy values). 

I had hoped that it would be possible to do this in a way that would break NengoOCL gracefully if somebody uses an unsupported SparseTransform (i.e. we'd get the nice "That operator is not supported" message that's currently implemented for unsupported ops), but I'm getting less and less sure if that's possible. So maybe we just do things in whatever way makes the most sense here, and then backends have to catch up. It would be nice, though, to have an easier way to check what type of signal something is than having to do e.g. `isinstance(sig.initial_value, scipy.sparse.spmatrix)` or something like that.

Reply:
Yeah, definitely agreed on all of that. I think in this case I just won't have a good sense of how feasible this is until I start trying it and see what breaks. But if trying to remove `SparseTransform` seems like it'll make things more complicated rather than less so, then I'm fine with just rolling back to this.

Reply:
> We make assumptions in a few places about how signals are done internally, and I worry that backends are making more assumptions.

One of the major places where this is the case, is probably the optimizer.